File: bindings/dotnet/samples/Pamoja.Boards/RaspberryPi/Arm.cs
// ANCHOR: example
using Pamoja.Actuators;
using Pamoja.Hal;
using Pamoja.Kit;

namespace Boards.RaspberryPi;

/// <summary>
/// The inspection rover's arm: a shoulder servo and an elbow servo on a PCA9685 board on the
/// header's I2C bus, reaching for the controls on an inverter cabinet's panel. Wire the PCA9685
/// board's SDA to GPIO2, SCL to GPIO3, VCC to 3V3, and GND to ground, with the shoulder servo on
/// channel 0, the elbow servo on channel 1, and V+ from a 5 V supply whose ground is joined to
/// the Pi's.
/// </summary>
public static class Arm
{
    // The PCA9685 channels the shoulder and elbow servos are plugged into.
    private const byte ShoulderChannel = 0;
    private const byte ElbowChannel = 1;

    // The panel's controls, each in meters out from the shoulder and up from it.
    private static readonly (string Name, float X, float Y)[] Panel =
    [
        ("reset button", 0.35f, 0.20f),
        ("breaker", 0.45f, 0.10f),
        ("door latch", 0.20f, 0.35f),
        ("fan switch", 0.70f, 0.00f),
    ];

    /// <summary>Reaches for each control on the panel, then parks the arm.</summary>
    public static void Run()
    {
        // The PCA9685 on the header's I2C bus, at 50 Hz for the servos.
        using I2cBus bus = I2cBus.Open("/dev/i2c-1");
        using var controller = new Pca9685(bus, Pca9685.DefaultAddress, frequencyHz: 50);

        // Links of 0.30 m and 0.25 m. The servos were fitted with both joints at 0, so a joint
        // angle of 0 is each servo's center, 90 degrees.
        var arm = new TwoLinkArm(0.30f, 0.25f);
        ServoMap servo = ServoMap.Standard;
        ushort Pulse(float joint) => servo.Pulse((float)(90.0 + joint * 180.0 / Math.PI));

        // Each control the arm can reach, it holds for two seconds; one it cannot, it skips
        // rather than drive a servo into its end stop.
        foreach ((string name, float x, float y) in Panel)
        {
            (float Shoulder, float Elbow)? joints = arm.JointsFor(x, y, Elbow.Up);
            if (joints is null)
            {
                Console.WriteLine($"{name,-12}  out of reach, skipped");
                continue;
            }

            (float shoulder, float elbow) = joints.Value;
            controller.SetChannel(ShoulderChannel, Pwm.Servo(Pulse(shoulder), 50));
            controller.SetChannel(ElbowChannel, Pwm.Servo(Pulse(elbow), 50));
            Console.WriteLine($"{name,-12}  shoulder {Pulse(shoulder)} us, elbow {Pulse(elbow)} us");
            Thread.Sleep(TimeSpan.FromSeconds(2));
        }

        // Back to the pose the arm was fitted in. The PCA9685 goes on sending both pulses after
        // the program exits, so the servos hold the arm there.
        controller.SetChannel(ShoulderChannel, Pwm.Servo(Pulse(0.0f), 50));
        controller.SetChannel(ElbowChannel, Pwm.Servo(Pulse(0.0f), 50));
        Console.WriteLine($"parked        both servos at {Pulse(0.0f)} us");
    }
}
// ANCHOR_END: example

File: examples/boards/raspberry-pi/src/bin/arm.rs
//! The inspection rover's arm: a shoulder servo and an elbow servo on a PCA9685 board on the
//! header's I2C bus, reaching for the controls on an inverter cabinet's panel.
//!
//! Wire the PCA9685 board's SDA to GPIO2, SCL to GPIO3, VCC to 3V3, and GND to ground, and turn
//! the I2C interface on. Plug the shoulder servo into channel 0 and the elbow servo into channel
//! 1, and feed the board's V+ from a 5 V supply whose ground is joined to the Pi's. Then run
//! `cargo run --release --bin arm`. See docs/guides/motion.md.

use std::error::Error;
use std::thread;
use std::time::Duration;

// ANCHOR: example
use pamoja_actuators::pca9685::{self, Pca9685, Pwm};
use pamoja_hal::bus::I2cBus;
use pamoja_kit::{Elbow, ServoMap, TwoLinkArm};

/// The PCA9685 channels the shoulder and elbow servos are plugged into.
const SHOULDER_CHANNEL: u8 = 0;
const ELBOW_CHANNEL: u8 = 1;

/// The panel's controls, each in meters out from the shoulder and up from it.
const PANEL: [(&str, f32, f32); 4] = [
    ("reset button", 0.35, 0.20),
    ("breaker", 0.45, 0.10),
    ("door latch", 0.20, 0.35),
    ("fan switch", 0.70, 0.00),
];

fn main() -> Result<(), Box<dyn Error>> {
    // The PCA9685 on the header's I2C bus, at 50 Hz for the servos.
    let bus = I2cBus::open("/dev/i2c-1")?;
    let mut controller =
        Pca9685::new(bus.clone(), pca9685::DEFAULT_I2C_ADDRESS, bus.delay()).with_frequency(50);

    // Links of 0.30 m and 0.25 m. The servos were fitted with both joints at 0, so a joint
    // angle of 0 is each servo's center, 90 degrees.
    let arm = TwoLinkArm::new(0.30, 0.25);
    let servo = ServoMap::standard();
    let pulse = |joint: f32| u32::from(servo.pulse(90.0 + joint.to_degrees()));

    // Each control the arm can reach, it holds for two seconds; one it cannot, it skips rather
    // than drive a servo into its end stop.
    for (name, x, y) in PANEL {
        let Some((shoulder, elbow)) = arm.joints_for(x, y, Elbow::Up) else {
            println!("{name:12}  out of reach, skipped");
            continue;
        };
        controller.set_channel(SHOULDER_CHANNEL, Pwm::servo(pulse(shoulder), 50))?;
        controller.set_channel(ELBOW_CHANNEL, Pwm::servo(pulse(elbow), 50))?;
        println!(
            "{name:12}  shoulder {} us, elbow {} us",
            pulse(shoulder),
            pulse(elbow)
        );
        thread::sleep(Duration::from_secs(2));
    }

    // Back to the pose the arm was fitted in. The PCA9685 goes on sending both pulses after
    // the program exits, so the servos hold the arm there.
    controller.set_channel(SHOULDER_CHANNEL, Pwm::servo(pulse(0.0), 50))?;
    controller.set_channel(ELBOW_CHANNEL, Pwm::servo(pulse(0.0), 50))?;
    println!("parked        both servos at {} us", pulse(0.0));
    Ok(())
}
// ANCHOR_END: example
